oecophylla/function/function.rule

```diff
@@ -201,12 +206,116 @@ rule humann2:
         expand(rules.function_humann2_combine_tables.output, sample=samples)
 
 
-rule function:
+rule function_shogun:
     """
-    Rule to do functional profiling of metagenomes.
+    Runs SHOGUN to construct functional profiles
+    """
+    input:
+        profile = taxonomy_dir + "{sample}/shogun/{sample}.profile.txt"
+    output:
+        genus_norm = func_dir + "{sample}/shogun/{sample}.genus.normalized.txt",
```

Comment:
This is a lot of outputs!

I guess we could pass `level` as a parameter in the same way we do for Bracken. What do you think?

Reply:
Well, I originally planned the same way, but then changed mind. Reason:

1) SHOGUN's taxonomic profiling and functional profiling have different available levels. For taxonomy, kingdom to strain are all available; but for function, only genus, species and strains are available.

2) For SHOGUN's functional profiling, there is no other output file except for the level-specific files. If I do the same mechanism as I did for taxonomic profiling (i.e., letting user specify level), there won't be any Snakemake output file specified.

Reply:
I'm not sure I understand point #2. 

Could it not work like:

```python
per_sample_rule:
  input:
    taxonomy_dir + "{sample}/shogun/{sample}.profile.txt"
  output:
    func_dir + "{sample}/shogun/{sample}.{level}.normalized.txt"

combine_rule:
  input:
    expand(func_dir + "{sample}/shogun/{sample}.{level}.normalized.txt", sample=samples)
  output:
    func_dir + "shogun/combined.{level}.normalized.txt"
```

The primary downside here is that it would be running a bunch of additional jobs, but these should be pretty lightweight, right? It's not like having to re-run Kraken a bunch of times?

Reply:
Yes, the functional profiling is very light. In your proposed solution, the program is able to generate only one level per run. Whereas in the taxonomy module, the user can specify any number of levels. Plus, I am not sure if this `{level}` syntax is valid in the `input`/`output` section. What do you think?

Reply:
This is a valid syntax. I've appended a complete toy example that I just ran. 

This syntax will allow you to use as many elements of {level} as you specify in a config. It will just allocate a new job tree for each level. The reason that I think this is preferable, in this case, is that 1) the small jobs have relatively light and homogenous resource requirements, and 2) it avoids flooding the directories with outputs. 

In fact, if these are light, the better solution might be just to run the whole thing as a single job per level? That way you could have all the per-sample file outputs be handled on local scratch and avoid a lot of writes to disk. Thoughts?

```
samples = ['sample1','sample2']
levels = ['level1','level2']

rule init:
    output:
        "{sample}.profile.txt"
    run:
        shell("echo {wildcards.sample} > {output}")

rule per_sample:
    input:
        "{sample}.profile.txt"
    output:
        "{sample}.{level}.normalized.txt"
    run:
        shell("""
                cat {input} > {output}
              """) 

rule combine:
    input:
        lambda wildcards: expand("{sample}.{level}.normalized.txt",
                                 sample = samples,
                                 level = wildcards.level)
    output:
        "combined.{level}.txt"
    run:
        shell("""
                cat {input} > {output}
              """)

rule all:
    input:
        expand(rules.combine.output, level = levels)
```

It produces the following outputs:

```
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample2.profile.txt
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample2.level2.normalized.txt
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample2.level1.normalized.txt
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample1.profile.txt
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample1.level2.normalized.txt
-rw-r--r--+ 1 jonsanders  staff    8 Oct 11 10:14 sample1.level1.normalized.txt
-rw-r--r--+ 1 jonsanders  staff   16 Oct 11 10:14 combined.level2.txt
-rw-r--r--+ 1 jonsanders  staff   16 Oct 11 10:14 combined.level1.txt
```

Reply:
@tanaes Your example sounds great. I would like to give it a try. Meanwhile, I don't fully understand your last paragraph. Do you mean merging"per_sample" and "combine" into one rule? I don't think there will be lots of disk writes. If only one level is specified, there will be four files generated.

Reply:
Meaning so if I specify three levels and those output three file types per sample and I have 3000 samples I end up with 27k files, but I really only care about the combined biom. So maybe worth doing shogun functions profiling from all?

Reply:
Meaning so if I specify three levels and those output three file types per sample and I have 3000 samples I end up with 27k files, but I really only care about the combined biom. So maybe worth doing shogun functions profiling from all?

Reply:
I am okay with it. But with 3000 samples you will get some 60k qc, raw, and taxonomy files anyway. Are you sure you don't want intermediates?

Reply:
Hi @tanaes I am looking into this old PR. Could you confirm that you really don't want per-sample functional profiles but just a study-wide one? Thanks!